### ISSUE TITLE
Not generating routes from controller correctly

### ISSUE THREAD

**Issue by devlinpadayachee** [theoomoregbee/sails-hook-swagger-generator#11]
Hi I get something along the lines of 

No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'
No tag for this identity 'v1/vdc'

When I sails lift. These are related to the functions in my controller. Is there anyway to have the swagger document generated without having to directly specify the routes again in config.routes.js 

like so

  ```
'get /v1/vdc/provisionNew': {
    controller: 'v1/vdcController',
    action: 'provisionNew',
    swagger: {
      summary: 'ProvisionNew',
      description: 'This is for authentication of any user',
      body: {
        username: { type: 'string', required: true },
        password: { type: 'password', required: true }
      },
      query: [{ $ref: '#/parameters/IDPathParam' }]
    }
  }
```

It would be better if it could autogenerate from the controller without having to redo the routes.

**Comment by theo4u** [comment]
Oh it generates from the controller, maybe it was unable to match the controller to any model, which is normal. But it can't detect if the controller is a `post` or `get` which it takes default to `get` and `GET` don't allow body.

**Comment by devlinpadayachee** [comment]
Hi,

My controller looks as follows

`/**
 * V1/vdcController
 *
 * @description :: Server-side logic for managing v1/vdcs
 * @help        :: See http://sailsjs.org/#!/documentation/concepts/Controllers
 */
module.exports = {
  provisionNew: async(req, res) => {
    try {
      var vdcAuthToken = await vdcService.getAuthToken();
      sails.log('Got vdcAuthToken: ' + vdcAuthToken);
    } catch(error) {
      sails.log(error.message);
      return res.ignitionError(error.message);
    }
    try {
      var vdc = await vdcService.createVdc(vdcAuthToken);
      sails.log(typeof vdc);
      sails.log(vdc);
    } catch(error) {
      sails.log(error);
      return res.ignitionError(error);
    }
    return res.ignitionMeta(vdc);
  },
  provisionView: async(req, res) => {
    if(req.param('vdcId') === undefined) {
      return res.ignitionError('No vdcId field present');
    }
    try {
      var vdcAuthToken = await vdcService.getAuthToken();
      sails.log('Got vdcAuthToken: ' + vdcAuthToken);
    } catch(error) {
      sails.log(error.message);
      return res.ignitionError(error.message);
    }
    try {
      var vdc = await vdcService.getVdc(vdcAuthToken, req.param('vdcId'));
    } catch(error) {
      return res.ignitionError(error);
    }
    return res.send(vdc);
  },
  provisionMigrate: async(req, res) => {
    if(req.param('vdcId') === undefined) {
      return res.ignitionError('No vdcId field present');
    }
    try {
      var vdcAuthToken = await vdcService.getAuthToken();
      sails.log('Got vdcAuthToken: ' + vdcAuthToken);
    } catch(error) {
      sails.log(error.message);
      return res.ignitionError(error.message);
    }
    try {
      var vdcMigrate = await vdcService.updateVdc(vdcAuthToken, req.param('vdcId'), {active: true});
      sails.log('Vdc Migrate: ' + JSON.stringify(vdcMigrate));
    } catch(error) {
      sails.log(error);
      return res.ignitionError(error);
    }
    return res.ok(vdcMigrate);
  },
  provisionSuspend: async(req, res) => {
    if(req.param('vdcId') === undefined) {
      return res.ignitionError('No vdcId field present');
    }
    try {
      var vdcAuthToken = await vdcService.getAuthToken();
      sails.log('Got vdcAuthToken: ' + vdcAuthToken);
    } catch(error) {
      sails.log(error.message);
      return res.ignitionError(error.message);
    }
    try {
      var vdcSuspend = await vdcService.updateVdc(vdcAuthToken, req.param('vdcId'), {active: true});
      sails.log('Vdc Suspend: ' + JSON.stringify(vdcSuspend));
    } catch(error) {
      sails.log(error);
      return res.ignitionError(error);
    }
    return res.ok(vdcSuspend);
  },
  provisionUnSuspend: async(req, res) => {
    if(req.param('vdcId') === undefined) {
      return res.ignitionError('No vdcId field present');
    }
    try {
      var vdcAuthToken = await vdcService.getAuthToken();
      sails.log('Got vdcAuthToken: ' + vdcAuthToken);
    } catch(error) {
      sails.log(error.message);
      return res.ignitionError(error.message);
    }
    try {
      var vdcUnSuspend = await vdcService.updateVdc(vdcAuthToken, req.param('vdcId'), {active: false});
      sails.log('Vdc Unsuspend: ' + JSON.stringify(vdcUnSuspend));
    } catch(error) {
      sails.log(error);
      return res.ignitionError(error);
    }
    return res.ok(vdcUnSuspend);
  },
  provisionDelete: async(req, res) => {
    if(req.param('vdcId') === undefined) {
      return res.ignitionError('No vdcId field present');
    }
    try {
      var vdcAuthToken = await vdcService.getAuthToken();
      sails.log('Got vdcAuthToken: ' + vdcAuthToken);
    } catch(error) {
      sails.log(error.message);
      return res.ignitionError(error.message);
    }
    try {
      var deletedVdc = await vdcService.deleteVdc(vdcAuthToken, req.param('vdcId'));
      sails.log('Deleted Vdc: ' + deletedVdc);
    } catch(error) {
      sails.log(error);
      return res.ignitionError(error);
    }
    return res.send(deletedVdc);
  },
  getLocations: async(req, res) => {
    try {
      var vdcAuthToken = await vdcService.getAuthToken();
      sails.log('Got vdcAuthToken: ' + vdcAuthToken);
    } catch(error) {
      sails.log(error.message);
      return res.ignitionError(error.message);
    }
    try {
      var locations = await vdcService.getLocations(vdcAuthToken);
      sails.log('Locations: ' + JSON.stringify(locations));
    } catch(error) {
      sails.log(error);
      return res.ignitionError(error);
    }
    return res.send(locations);
  }
};`

**Comment by devlinpadayachee** [comment]
So for each of the controller functions above, I have to redo the same setup in the routes.js to get the swagger to work correctly

'get /v1/vdc/provisionNew': {
  controller: 'v1/vdcController',
  action: 'provisionNew',
  swagger: {
    summary: 'ProvisionNew',
    description: 'This is for authentication of any user',
    body: {
      username: { type: 'string', required: true },
      password: { type: 'password', required: true }
    },
    query: [{ $ref: '#/parameters/IDPathParam' }]
  }
}

**Comment by theo4u** [comment]
Oh i guess the error, was that your control name is containing `/`

**Comment by devlinpadayachee** [comment]
Hi is there a way to do it with the /v1/{controller}/{method} as i use the /v1/ for version control

**Comment by theo4u** [comment]
oh yeaa, check rest prefix and prefix https://sailsjs.com/documentation/reference/configuration/sails-config-blueprints

**Comment by rvmey** [comment]
I have this same problem.  
I can't seem to get the blueprint paths to include /api/{controller}/{method}

I've tried various combinations of these settings in blueprints.js:
module.exports.blueprints = {
  rest: true,
  shortcuts: true,
  restPrefix: '/api',
  prefix: '/api',

My swagger endpoints still end up with /.  
curl -X GET "http://localhost:1337/" -H "accept: application/json"

I would expect this result because my controller name is cloud.  
curl -X GET "http://localhost:1337/api/cloud" -H "accept: application/json"

I'm using the default generated controller file:

/**
 * CloudController
 *
 * @description :: Server-side actions for handling incoming requests.
 * @help        :: See https://sailsjs.com/docs/concepts/actions
 */
module.exports = { 
};

**Comment by theo4u** [comment]
Which version of sailjs are you using @rvmey

**Comment by theo4u** [comment]
> Is there a way to prevent certain routes from showing up in the swagger UI?

No, currently no way to specify if a route should be excluded

> Is there a way to have your blueprint routes show up without having to add them to routes.js

It's already working like this, as far your config.blueprint is `true`

**Comment by rvmey** [comment]
Can you please be specific about how to set config.blueprint = true?

I've tried these two settings based on #37.

module.exports.blueprints = {
  rest: true,
  shortcuts: true,

**Comment by theo4u** [comment]
> rest: true 

Should do it, then auto-generated routes for each model controller. Look at the comment here https://github.com/theo4u/sails-hook-swagger-generator/blob/a5e58a645333d5c8b5e1026acc83de834f1888a6/config/blueprints.js#L49

**Comment by rvmey** [comment]
Theo, I think you're telling me how to enable the blueprint rest endpoints.  I knew how to do that.  

What I was hoping your sails-hook-swagger-generator project would do is automatically give me swagger entries for each of those blueprints without having to manually add them to routes.js.  

I think that's what devlinpadayachee was asking for too when he wrote, "for each of the controller functions above, I have to redo the same setup in the routes.js to get the swagger to work correctly"

If sails-hook-swagger-generator doesn't do that automatically, then no problem, I'll do it manually.

**Comment by rvmey** [comment]
I hate to give up, but I can't get it to work.  
If you're interested, you could try my repo:  https://github.com/rvmey/SailsSwaggerTest

**Comment by theo4u** [comment]
cool, lemme take a look

**Comment by rvmey** [comment]
Good find.  Any ideas?  I think this is beyond me.

**Comment by theo4u** [comment]
Oh shit, thanks @xming13, do you want to take on it and use https://github.com/rvmey/SailsSwaggerTest as a test repo. Lemme know @xming13

**Comment by xming13** [comment]
@theo4u I tried some quick work around but did not manage to get it working. Seems like there is a need to rewrite based on the changes to sails v1.0. 
I only started using sails this week and I am afraid I am not able to help on this.

I will share what I have tried below:

I have `sails-hook-custom-blueprints` module in my repo. 
I am able to retrieve the actions generated by blueprints via `sails.getActions()` in  `hook:blueprints:loaded` event. (`sails-hook-custom-blueprints` module uses `hook:blueprints:loaded` event to register action and since there is no guarantee on the order of execution, there might be a better way to do this)

`sails-hook-swagger-generator/index.js`
```
       // Run when sails loads-- be sure and call `next()`.
        initialize: function (next) {
          var hook = this;

          // After the blueprints hook loads, 
          sails.after('hook:blueprints:loaded', function() {
            console.log(sails.getActions(), '[hook:blueprints:loaded] sails-hook-swagger-generator');
            return swaggerDoc(sails, hook, next);
          });
        }
```

**Comment by theo4u** [comment]
ok cool, thanks. Will just a create an issue for a rewrite for version 1

**Comment by vbarzana** [comment]
Same error here still, seems you guys didn't find a solution, I am using sails 0.13, any idea?
![image](https://user-images.githubusercontent.com/3055285/60699014-b0660780-9efa-11e9-9c3f-0beb7a7e0d3e.png)

**Comment by theoomoregbee** [comment]
fixed in #58
